Vega WebGPU Extension Implemented

KaNaDaAT opened the pull request:
I am not yet happy with everything but over all there now exists a Vega WebGPU Extension

KaNaDaAT committed:
fixing problems

KaNaDaAT committed:
ignroe build

KaNaDaAT committed:
Now import via link directly and minor fixes

KaNaDaAT committed:
feat: Background color now white!

KaNaDaAT committed:
fixed rule drawing

KaNaDaAT committed:
feat: text rendering and error fix on hover

KaNaDaAT committed:
feat: Rules now work

KaNaDaAT committed:
fix: general improvement

KaNaDaAT committed:
fix: Hover effect didnt work with div

KaNaDaAT committed:
update: packages updated

Updated packages and removed critical ones

KaNaDaAT committed:
feat: VertexBufferManager

VertexBufferManager for simpler vertext creation

KaNaDaAT committed:
feat: Path/Symbol start

KaNaDaAT committed:
feat: cleanup with custom methods. Symbol WORK!

KaNaDaAT committed:
refactor: Removed unnecessary files

KaNaDaAT committed:
feat: arc

KaNaDaAT committed:
fix: Removed debug color

KaNaDaAT committed:
feat: area

KaNaDaAT committed:
feat: area dirty fix

KaNaDaAT committed:
fix: depth texture must be recreated on size changes

KaNaDaAT committed:
fix: some default values

KaNaDaAT committed:
feat: color updata and rect support borders

KaNaDaAT committed:
fix: rule

KaNaDaAT committed:
todo: line

KaNaDaAT committed:
fix: line

KaNaDaAT committed:
update: General improvments

line improvment
rect bugfix with strokes

KaNaDaAT committed:
feat: symbols now can have strokes

KaNaDaAT committed:
feat: render rework (less code duplication)

KaNaDaAT committed:
fix: shorter render code for line, arc and area as well

KaNaDaAT committed:
feat: IT WORKS!

KaNaDaAT committed:
feat: performance update

KaNaDaAT committed:
feat: path now renders strokes correctly

KaNaDaAT committed:
fix: arc

KaNaDaAT committed:
feat: more improvements

KaNaDaAT committed:
fix: location of paths

KaNaDaAT committed:
fix: smaller issues fixed and line performance

KaNaDaAT committed:
feat: Performance update.

KaNaDaAT committed:
Release: 0.1

KaNaDaAT committed:
small rebuild for test

KaNaDaAT committed:
Version: 1.0.0

KaNaDaAT committed:
Update index.html

KaNaDaAT committed:
Release: Version 1.0.0

KaNaDaAT commented:
Readme might need some changes for referenzing your hosts

KaNaDaAT committed:
Update README.md

KaNaDaAT committed:
Dev Stuff

KaNaDaAT committed:
Squashed commit of the following:

commit c229507f5810a207e8da5158876adee3d0587e42
Author: KaNaDaAT <kalian.danzer@gmail.com>
Date:   Sun Jan 14 16:49:21 2024 +0100

    Release: Version 1.1.0

    Render Bundles

KaNaDaAT committed:
Update: Demo updated to 1.1.0

KaNaDaAT committed:
Feat: SImple Release

KaNaDaAT committed:
1_0_0 currently bugged

lsh commented on src/util/renderer.ts:
```diff
@@ -0,0 +1,352 @@
+		});
+	}
+
+	static createPipelineLayout(name: string, device: GPUDevice, bindGroupLayouts: GPUBindGroupLayout[]) {
```
This was a mistake in my original implementation. We should not be creating a pipeline for each mark type, since that has major performance costs. We should instead try to create a single pipeline that uses [overrides](https://www.w3.org/TR/WGSL/#override-decls) to handle each mark type.

lsh commented on src/marks/line.ts:
```diff
@@ -1,219 +1,127 @@
-  return null;
+function createAttributes(items: SceneItem[]) {
+  const lines = items as SceneLine[];
+  const result = new Float32Array((items.length - 1) * 9);
```
We should be storing `9` as a constant where we define the vertex descriptor rather than as a magic number. Same for all other marks.

lsh commented on src/WebGPURenderer.ts:
```diff
@@ -11,27 +26,51 @@ export default function WebGPURenderer(loader: unknown) {
 }
 
-const base = Renderer.prototype;
+let base = Renderer.prototype;
```
This reminds me to convert Renderer to a class rather than just taking the prototype.

KaNaDaAT committed:
README.md aktualisieren

KaNaDaAT committed:
feat: Example supporting Versions now

KaNaDaAT committed:
Performance and structure changes

KaNaDaAT committed:
Release: 1.1.0

KaNaDaAT committed:
Update README.md

KaNaDaAT committed:
Example: Now can also use webgl

KaNaDaAT committed:
Merge branch 'main' of https://github.com/KaNaDaAT/vega-webgpu

KaNaDaAT committed:
Performance for path

KaNaDaAT committed:
Update benchmark.vg.json

KaNaDaAT committed:
Release: 1.1.0

KaNaDaAT committed:
update dependencies

KaNaDaAT committed:
Render loop improvements

When the last render has not finished yet it will store the most recent render call and run that at the end of the frame so it does not waste gpu time. Added option for it as well.

KaNaDaAT committed:
Rerelease: 1.1.1

KaNaDaAT committed:
Create .gitattributes

KaNaDaAT committed:
Update .gitattributes

KaNaDaAT committed:
Update README.md

KaNaDaAT committed:
Update README.md

KaNaDaAT committed:
feat: clipping, line performance,

KaNaDaAT committed:
fix: path offset

KaNaDaAT committed:
fix: renderBatch

KaNaDaAT committed:
Release: 1.2.0

Performance improvements on Lines.
Render Bundling Option `renderBatch`.
Now supports Clipping.
Fixed a bug regarding Path offsets.

KaNaDaAT committed:
Update README.md

domoritz commented:
Thanks for the update. @lsh, could you take a look?

KaNaDaAT commented:
Its still work in progess. I will actually close this PR and reopem once I am actually happy with it. 
Sorry if any notifications appeared.
I had a bit of a struggle orchestering my changes to somewhat good commits 😅
Maybe I should just have given up on that and done one big commit to be honest.
Still strugeling with the CI and some features are missing still.
